src/main/java/org/primeframework/mvc/security/cookiesession/UserIDCookieSessionSecurityContext.java

```diff
@@ -0,0 +1,236 @@
+ */
+package org.primeframework.mvc.security.cookiesession;
+
+import javax.crypto.BadPaddingException;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.ZonedDateTime;
+import java.util.Optional;
+import java.util.UUID;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.fusionauth.http.Cookie;
+import io.fusionauth.http.server.HTTPRequest;
+import io.fusionauth.http.server.HTTPResponse;
+import org.primeframework.mvc.ErrorException;
+import org.primeframework.mvc.security.CookieProxy;
+import org.primeframework.mvc.security.Encryptor;
+import org.primeframework.mvc.security.UnauthenticatedException;
+import org.primeframework.mvc.security.UserLoginSecurityContext;
+import org.primeframework.mvc.util.CookieTools;
+
+/**
+ * Logs in and out users based on storing their User ID in a cookie and also adds a sliding session
+ * timeout with a maximum age
+ *
+ */
+public abstract class UserIDCookieSessionSecurityContext implements UserLoginSecurityContext {
+  public static final String UserKey = "primeCurrentUser";
+
+  private final HTTPRequest request;
+
+  private final HTTPResponse response;
+
+  private final Encryptor encryptor;
+
+  private final ObjectMapper objectMapper;
+
+  protected final CookieProxy sessionCookie;
+
+  private final Clock clock;
+
+  private final Duration sessionMaxAge;
+
+  private final Duration sessionTimeout;
+
+  protected UserIDCookieSessionSecurityContext(HTTPRequest request,
+                                               HTTPResponse response,
+                                               Encryptor encryptor,
+                                               ObjectMapper objectMapper,
+                                               Clock clock,
+                                               Duration sessionTimeout,
+                                               Duration sessionMaxAge) {
+    this.request = request;
+    this.response = response;
+    this.encryptor = encryptor;
+    this.objectMapper = objectMapper;
+    this.clock = clock;
+    this.sessionMaxAge = sessionMaxAge;
+    this.sessionTimeout = sessionTimeout;
+    long timeoutInSeconds = sessionTimeout.toSeconds();
+    this.sessionCookie = new CookieProxy(getCookieName(), timeoutInSeconds, Cookie.SameSite.Strict);
+  }
+
+  protected String getCookieName() {
+    return UserKey;
+  }
+
+  enum CookieExtendResult {
+    // no extension needed
+    Keep,
+    // extend the cookie with the same value
+    Extend,
+    // the cookie is now invalid
+    Invalid
+  }
+
+  /**
+   * Figure out whether to extend cookie
+   *
+   * @param now            the time now
+   * @param signInTime     when user signed in originally
+   * @param maxSessionAge  max absolute session length
+   * @param sessionTimeout max session timeout
+   * @return correct cookie behavior
+   */
+  static CookieExtendResult shouldExtendCookie(ZonedDateTime now, ZonedDateTime signInTime, Duration maxSessionAge, Duration sessionTimeout) {
```

Comment:
Doesn't seem like this should be static. If it is only intended for use by this "context" object, seems like it should be `protected`?